File: src/benchmark/execution-lean.mjs
// Native artifact normalization for the private execution contract. Candidate
// logs cannot supply fill evidence; every reconciled receipt must match LEAN's
// own retained orders and OrderEvents. The caller supplies the reviewed intents.
import { canonical, invariant, object } from './prompts.mjs'
import { replayExecutionLedger } from './execution.mjs'

export const EXECUTION_TAG_VERSION = 'LB-EXEC-1'
const statusName = value => ({ 0: 'new', 1: 'submitted', 2: 'partiallyfilled', 3: 'filled', 5: 'canceled', 6: 'none', 7: 'invalid', 8: 'cancelpending', 9: 'updatesubmitted' }[value] || String(value).toLowerCase().replace(/[_ ]/g, ''))
const integer = value => Number.isSafeInteger(value) && value >= 0
export function executionCents(value) {
  const match = /^(\d+)(?:\.(\d+))?(?:[eE]([+-]?\d+))?$/.exec(String(value))
  invariant(match, 'Native execution amount is not nonnegative decimal cents.')
  const exponent = Number(match[3] || 0) + 2 - (match[2] || '').length
  invariant(Number.isSafeInteger(exponent) && Math.abs(exponent) <= 100, 'Native execution amount exceeds the exact range.')
  const digits = BigInt(match[1] + (match[2] || '')), scale = 10n ** BigInt(Math.abs(exponent))
  invariant(exponent >= 0 || digits % scale === 0n, 'Native execution amount has sub-cent precision.')
  const cents = exponent >= 0 ? digits * scale : digits / scale
  invariant(cents <= BigInt(Number.MAX_SAFE_INTEGER), 'Native execution amount exceeds the exact range.')
  return Number(cents)
}

export function normalizeExecutionOrders(orders, intents) {
  invariant(object(orders) && Array.isArray(intents), 'Retain native orders and declared intents.')
  const declared = new Map(intents.map(intent => [intent.id, intent])), seen = new Set(), result = new Map()
  invariant(declared.size === intents.length, 'Declared intent identifiers repeat.')
  for (const order of Object.values(orders)) {
    const tag = JSON.parse(order.tag ?? order.Tag ?? 'null')
    invariant(Array.isArray(tag) && tag.length === 2 && tag[0] === EXECUTION_TAG_VERSION, 'A native order has no execution tag.')
    const intent = declared.get(tag[1]), id = order.id ?? order.Id
    invariant(intent && !seen.has(intent.id) && integer(id) && !result.has(String(id)), 'Native orders do not bind one-to-one to declared intents.')
    const symbol = order.symbol?.value ?? order.Symbol?.Value
    const signed = intent.quantity * (intent.side === 'buy' ? 1 : -1)
    invariant(symbol === intent.asset && Number(order.quantity ?? order.Quantity) === signed, 'A native order changed its declared asset or quantity.')
    invariant(Number(order.type ?? order.Type) === 0, 'The execution adapter requires native market orders.')
    const time = Date.parse(order.time ?? order.Time) / 1000
    invariant(integer(time) && time === intent.time, 'Native order creation differs from the intent time.')
    result.set(String(id), { intent, status: statusName(order.status ?? order.Status) }); seen.add(intent.id)
  }
  return result
}

export function normalizeExecutionEvents(events, orders, intents, currency) {
  invariant(Array.isArray(events), 'Retain native execution events.')
  const native = normalizeExecutionOrders(orders, intents), result = [], seen = new Set()
  let previous = -1
  for (const event of events) {
    const orderId = event.orderId ?? event.OrderId, eventId = event.orderEventId ?? event.OrderEventId
    const binding = native.get(String(orderId))
    invariant(binding && integer(orderId) && integer(eventId), 'A native event has no bound order or event identity.')
    const { intent } = binding, key = String(orderId) + ':' + eventId
    invariant(!seen.has(key), 'Native event identities repeat.'); seen.add(key)
    const timeValue = event.time ?? event.Time ?? event.utcTime ?? event.UtcTime
    const time = typeof timeValue === 'number' ? timeValue : Date.parse(timeValue) / 1000
    invariant(integer(time) && time >= previous && time >= intent.time, 'Native execution chronology is invalid.'); previous = time
    invariant((event.symbolValue ?? event.SymbolValue) === intent.asset && !event.isAssignment && !event.IsAssignment, 'Native fill asset or assignment differs from this contract.')
    const status = statusName(event.status ?? event.Status)
    const kind = ({ submitted: 'accepted', partiallyfilled: 'fill', filled: 'fill', canceled: 'cancelled', cancelled: 'cancelled', invalid: 'rejected' })[status]
    const quantity = Number(event.fillQuantity ?? event.FillQuantity), fee = event.orderFeeAmount ?? event.OrderFeeAmount
    if (kind !== 'fill') {
      invariant(quantity === 0 && (fee === undefined || executionCents(fee) === 0), 'A non-fill event carries unaccounted inventory or fees.')
      invariant(kind || ['new', 'none', 'cancelpending'].includes(status), 'Unsupported native execution status.')
      if (!kind) continue
    }
    const row = { id: key, intentId: intent.id, brokerOrderId: String(orderId), kind, time }
    if (kind === 'fill') {
      invariant(Number.isSafeInteger(quantity) && quantity !== 0 && (quantity > 0) === (intent.side === 'buy'), 'A native fill has an invalid quantity or side.')
      invariant((event.fillPriceCurrency ?? event.FillPriceCurrency) === currency && (event.orderFeeCurrency ?? event.OrderFeeCurrency) === currency, 'Native execution currency differs.')
      invariant(fee !== undefined, 'A native fill omitted its fee receipt.')
      Object.assign(row, { quantity: Math.abs(quantity), priceCents: executionCents(event.fillPrice ?? event.FillPrice), feeCents: executionCents(fee) })
    }
    result.push(row)
  }
  return result
}

export function verifyNativeExecution({ config, result, events, journal, actions }) {
  const state = result.state ?? result.State, orders = result.orders ?? result.Orders
  invariant(state && (state.Status ?? state.status) === 'Completed' && !(state.RuntimeError || state.runtimeError || result.RuntimeError || result.runtimeError), 'Native execution did not complete without a runtime error.')
  const declared = actions.filter(action => action.kind === 'intent').map(action => action.intent)
  const normalized = normalizeExecutionEvents(events, orders, declared, config.currency)
  const retained = journal.filter(row => row.kind === 'broker-event').map(row => { const { sequence, ...event } = row.event; return event })
  invariant(canonical(retained) === canonical(normalized), 'The execution journal does not match the complete native receipt history.')
  const decisions = journal.filter(row => row.kind !== 'broker-event').map(row => row.kind === 'intent'
    ? { kind: 'intent', intent: row.decision.intent } : { kind: 'cancel', intentId: row.event.intentId, time: row.event.time })
  invariant(canonical(decisions) === canonical(actions), 'The execution journal changed the declared action history.')
  const snapshot = replayExecutionLedger(config, journal).snapshot(), native = normalizeExecutionOrders(orders, declared)
  invariant(native.size === snapshot.tickets.length, 'Native orders and admitted private tickets differ.')
  const final = { submitted: 'accepted', partiallyfilled: 'partially-filled', filled: 'filled', canceled: 'cancelled', cancelled: 'cancelled', invalid: 'rejected' }
  for (const ticket of snapshot.tickets) {
    const binding = native.get(ticket.brokerOrderId)
    invariant(binding?.intent.id === ticket.id && (final[binding.status] === ticket.status || binding.status === 'cancelpending' && ticket.cancellationRequested && ['accepted', 'partially-filled'].includes(ticket.status)), 'A private ticket differs from the terminal native order state.')
  }
  // Check every partial/full status, including intermediate events.
  const filled = new Map()
  for (const event of events) {
    const status = statusName(event.status ?? event.Status)
    if (!['filled', 'partiallyfilled'].includes(status)) continue
    const id = String(event.orderId ?? event.OrderId), binding = native.get(id)
    const total = (filled.get(id) || 0) + Math.abs(Number(event.fillQuantity ?? event.FillQuantity)); filled.set(id, total)
    invariant(status === 'filled' ? total === binding.intent.quantity : total < binding.intent.quantity, 'Native fill status disagrees with its accumulated quantity.')
  }
  return snapshot
}

File: src/benchmark/execution_lean.py
"""LEAN market-order adapter for the explicit private execution ledger.

Admission precedes dispatch. Only real OrderEvent callbacks settle inventory
and cash. A cancellation request does not acknowledge its own success.
The caller selects/reviews trading semantics and the backtest fill model.
"""
import json
from datetime import timezone
from decimal import Decimal
from AlgorithmImports import OrderStatus
from execution_reference import require, checked

TAG_VERSION = "LB-EXEC-1"


def cents(value):
    result = Decimal(str(value)) * 100
    require(result == result.to_integral_value(), "Native amount is not exact cents")
    return checked(int(result))


def seconds(value):
    result = value.replace(tzinfo=timezone.utc).timestamp()
    require(result == int(result), "Native time is not an integer UTC second")
    return checked(int(result))


class LeanExecutionBridge:
    def __init__(self, algorithm, ledger, symbols, reconciler=None, tagger=None):
        require(set(symbols) == set(ledger.assets), "Native asset registry differs")
        self.algorithm, self.ledger, self.symbols = algorithm, ledger, dict(symbols)
        self.reconcile = reconciler or ledger.reconcile
        self.tagger = tagger or (lambda intent: [TAG_VERSION, intent["id"]])
        self.tickets, self.raw_events = {}, []
        for asset, symbol in self.symbols.items():
            security = algorithm.securities[symbol]
            require(security.quote_currency.symbol == ledger.config["currency"], "Native quote currency differs")
            require(security.symbol_properties.contract_multiplier == ledger.assets[asset]["multiplier"], "Native asset multiplier differs")

    def submit(self, intent):
        require(intent["time"] == seconds(self.algorithm.utc_time), "Intent timestamp differs from dispatch")
        decision = self.ledger.admit(intent)
        if not decision["admitted"]:
            return decision
        self.dispatch(intent)
        return decision

    def dispatch(self, intent):
        require(intent["time"] == seconds(self.algorithm.utc_time), "Intent timestamp differs from dispatch")
        admitted = self.ledger.orders.get(intent["id"])
        require(admitted is not None and all(admitted.get(key) == value for key, value in intent.items()), "Native dispatch differs from its admitted intent")
        require(intent["id"] not in self.tickets and admitted["brokerOrderId"] is None, "Native intent was already dispatched")
        quantity = intent["quantity"] * (1 if intent["side"] == "buy" else -1)
        ticket = self.algorithm.market_order(self.symbols[intent["asset"]], quantity, True,
                                            tag=json.dumps(self.tagger(intent), separators=(",", ":")))
        # A callback may have settled this order before market_order returns.
        # Do not fabricate acceptance or replace its reconciled status.
        order = self.ledger.orders[intent["id"]]
        require(order["brokerOrderId"] in (None, str(ticket.order_id)), "Native ticket identity differs from its callback")
        self.tickets[intent["id"]] = ticket
        return ticket

    def cancel(self, intent_id):
        if not self.ledger.request_cancel(intent_id, seconds(self.algorithm.utc_time)):
            return False
        return self.dispatch_cancel(intent_id)

    def dispatch_cancel(self, intent_id):
        require(self.ledger.orders.get(intent_id, {}).get("cancellationRequested"), "Native cancellation needs its recorded request")
        require(intent_id in self.tickets, "No native ticket for cancellation")
        response = self.tickets[intent_id].cancel()
        # A failed request remains a request. Release reservations only when
        # a terminal broker event arrives.
        return bool(response.is_success)

    def on_order_event(self, event):
        order = self.algorithm.transactions.get_order_by_id(event.order_id)
        require(order is not None, "Native event has no order")
        tag = json.loads(order.tag)
        require(isinstance(tag, list) and len(tag) > 1 and isinstance(tag[-1], str), "Native order has no execution tag")
        intent = self.ledger.orders.get(tag[-1])
        require(intent is not None and tag == self.tagger(intent), "Native order has no admitted intent or changed its attribution")
        require(event.symbol == self.symbols[intent["asset"]], "Native event changed its asset")
        signed = intent["quantity"] * (1 if intent["side"] == "buy" else -1)
        require(order.quantity == signed, "Native order changed its admitted quantity")
        raw = dict(orderId=int(event.order_id), orderEventId=int(event.id), status=str(event.status),
                   time=seconds(event.utc_time), fillQuantity=str(event.fill_quantity), fillPrice=str(event.fill_price),
                   feeAmount=str(event.order_fee.value.amount), feeCurrency=event.order_fee.value.currency)
        self.raw_events.append(raw)
        kind = {OrderStatus.SUBMITTED: "accepted", OrderStatus.PARTIALLY_FILLED: "fill", OrderStatus.FILLED: "fill",
                OrderStatus.CANCELED: "cancelled", OrderStatus.INVALID: "rejected"}.get(event.status)
        if kind is None:
            require(event.status in (OrderStatus.NEW, OrderStatus.NONE, OrderStatus.CANCEL_PENDING), "Unsupported native order status")
            require(event.fill_quantity == 0 and event.order_fee.value.amount == 0, "Unreconciled native cash or fill")
            return
        normalized = dict(id=str(event.order_id) + ":" + str(event.id), intentId=intent["id"],
                          brokerOrderId=str(event.order_id), kind=kind, time=raw["time"])
        if kind == "fill":
            quantity = abs(event.fill_quantity)
            require(quantity == int(quantity) and quantity > 0 and (event.fill_quantity > 0) == (intent["side"] == "buy"), "Unsupported native fill quantity")
            require(event.order_fee.value.currency == self.ledger.config["currency"], "Native fee currency differs")
            normalized.update(quantity=int(quantity), priceCents=cents(event.fill_price), feeCents=cents(event.order_fee.value.amount))
        else:
            require(event.fill_quantity == 0 and event.order_fee.value.amount == 0, "Non-fill native event carries a fill or fee")
        self.reconcile(normalized)
        expected_status = "filled" if event.status == OrderStatus.FILLED else "partially-filled" if event.status == OrderStatus.PARTIALLY_FILLED else None
        require(expected_status is None or self.ledger.orders[intent["id"]]["status"] == expected_status, "Native fill status differs from its accumulated quantity")

    def verify_portfolio(self):
        require(cents(self.algorithm.portfolio.cash) == self.ledger.cash, "Native and private cash differ")
        snapshot = self.ledger.snapshot()
        for position in snapshot["positions"]:
            require(self.algorithm.portfolio[self.symbols[position["asset"]]].quantity == position["quantity"], "Native and private inventory differ")
        return snapshot
